service/src/main/java/io/techery/janet/AsyncActionService.java

```diff
@@ -168,47 +193,52 @@ private void connect(boolean reconnectIfConnected) throws AsyncServiceException
         }
     }
 
-    private void disconnect() throws AsyncServiceException {
+    private void onMessageReceived(Message message) {
+        Throwable extractError = null;
+        BytesArrayBody payloadBody = null;
         try {
-            client.disconnect();
-        } catch (Throwable t) {
-            throw new AsyncServiceException(t);
-        }
-    }
-
-    private void onMessageReceived(String event, BytesArrayBody body) {
-        if (!actionsRoster.containsEvent(event)) {
-            System.err.println(String.format("Received sync message %s is not defined by any action :(. The message contains body %s", event, body));
-            return;
+            payloadBody = extractPayload(message);
+        } catch (Throwable throwable) {
+            extractError = throwable;
         }
-        List<Class> actionClassList = actionsRoster.getActionClasses(event);
-        for (Class actionClass : actionClassList) {
+        final IncomingMessage incomingMessage = new IncomingMessage(message, payloadBody, converter);
+        List<Class> actionClasses = actionsRoster.getActionClasses(message.getEvent());
+        for (Class actionClass : actionClasses) {
             ActionHolder holder = ActionHolder.create((createActionInstance(actionClass)));
+            if (extractError != null) {
+                callback.onFail(holder, new AsyncServiceException(extractError));
+                continue;
+            }
             AsyncActionWrapper actionWrapper = getAsyncActionWrapper(holder);
             try {
-                actionWrapper.fillPayload(body, converter);
+                actionWrapper.setPayload(incomingMessage.getPayloadAs(actionWrapper.getPayloadFieldType()));
             } catch (ConverterException e) {
                 callback.onFail(holder, new AsyncServiceException(e));
+                continue;
             }
-            if (synchronizer.contains(event)) {
-                for (AsyncActionWrapper wrapper : synchronizer.sync(event, actionWrapper.action, new AsyncActionSynchronizer.Predicate() {
-                    @Override public boolean call(AsyncActionWrapper wrapper, Object responseAction) {
-                        try {
-                            return wrapper.fillResponse(responseAction);
-                        } catch (ConverterException e) {
-                            callback.onFail(wrapper.holder, new AsyncServiceException(e));
-                        }
-                        return false;
-                    }
-                })) {
-                    callback.onSuccess(wrapper.holder);
+            callback.onSuccess(holder);
+        }
+        for (AsyncActionWrapper wrapper : synchronizer.sync(new AsyncActionSynchronizer.Callback() {
```

Comment:
would like to extract into two more methods: `handleMessageAsIncoming` and `handleMessageAsPending`